### PR TITLE
Allow to optionally use C++17

### PR DESCRIPTION
**Pull request by MarcelKoch**
Based on: #1544 

This PR allows to optionally use C++17. If the CMake option `CMAKE_CXX_STANDARD` is set, this will be the standard that is used to compile ginkgo. If this option is not used, then the highest available standard between `c++14, c++17` will be used.

Additionally, the macro `GINKGO_HAVE_CXX17` will be set to true or false, depending on the standard.

Exception: DPCPP will still use C++17, and I have no idea of how the HIP setup works, but I'm guessing that it will ignore this.

PR Stack:

- [ ] #1545 
- [ ] #1543
- [ ] #1579
- [ ] #1544 
- [ ] #1548 
- [ ] #1546 
- [ ] #1549

### CODE REVIEW

File: cmake/cxx_std_check.cpp [hunk 1]
```diff
@@ -0,0 +1,5 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+int main() { static_assert(__cplusplus >= CXX_STD); }
```
Comment (thoasm) [inline cmake/cxx_std_check.cpp:5]: [`static_assert`](https://en.cppreference.com/w/cpp/language/static_assert) should also have an error string (unless it's C++17+):
```suggestion
int main() { static_assert(__cplusplus >= CXX_STD, "Required C++ standard not supported."); }
```

File: cmake/create_test.cmake [hunk 2]
```diff
@@ -32,7 +32,6 @@ function(ginkgo_set_test_target_properties test_target_name test_library_suffix)
     else()
         target_link_libraries(${test_target_name} PRIVATE ginkgo_gtest_main${test_library_suffix})
     endif()
-    target_compile_features(${test_target_name} PUBLIC cxx_std_14)
```
Comment (pratikvn) [inline cmake/create_test.cmake:35]: Is a setting to `${GINKGO_CXX_STANDARD_FEATURE}` not necessary here ? Same question for the test cmake files below.

File: include/ginkgo/config.hpp.in [hunk 3]
```diff
@@ -112,4 +112,10 @@
 // clang-format on
 
 
+/* Do we support C++17? */
```
Comment (pratikvn) [inline include/ginkgo/config.hpp.in:115]: minor nit: 
```suggestion
/* Does the CXX compiler support C++17? */
```

**Review by pratikvn — APPROVED** [review]
LGTM! I think we need to note that this will might break the OpenCARP pipelines ? Because they want to explicitly set to C++14. It is not an issue, but probably needs an update from their side to set the `GINKGO_CXX_STANDARD`.

File: cmake/create_test.cmake (hunk 2 above)
Comment (MarcelKoch) [reply cmake/create_test.cmake:35]: I think it was unnecessary to begin with. The test is linked against ginkgo, so the standard requirement from there is propagated.

**Comment by MarcelKoch** [comment]
This PR is not required anymore. If we decide at a later point to enable c++17 optionally we can revive this.

**Pull request closed without merging**